File: src/factories/StatementFactory.ts
import ts from "typescript";

export namespace StatementFactory {
    export const mut = (name: string, initializer?: ts.Expression) =>
        ts.factory.createVariableStatement(
            undefined,
            ts.factory.createVariableDeclarationList(
                [
                    ts.factory.createVariableDeclaration(
                        name,
                        undefined,
                        undefined,
                        initializer,
                    ),
                ],
                ts.NodeFlags.Let,
            ),
        );

    export const constant = (name: string, initializer?: ts.Expression) =>
        ts.factory.createVariableStatement(
            undefined,
            ts.factory.createVariableDeclarationList(
                [
                    ts.factory.createVariableDeclaration(
                        name,
                        undefined,
                        undefined,
                        initializer,
                    ),
                ],
                ts.NodeFlags.Const,
            ),
        );

    export const entry = (key: string) => (value: string) =>
        ts.factory.createVariableDeclarationList(
            [
                ts.factory.createVariableDeclaration(
                    ts.factory.createArrayBindingPattern([
                        ts.factory.createBindingElement(
                            undefined,
                            undefined,
                            ts.factory.createIdentifier(key),
                            undefined,
                        ),
                        ts.factory.createBindingElement(
                            undefined,
                            undefined,
                            ts.factory.createIdentifier(value),
                            undefined,
                        ),
                    ]),
                    undefined,
                    undefined,
                    undefined,
                ),
            ],
            ts.NodeFlags.Const,
        );

    export const transpile = (script: string) =>
        ts.factory.createExpressionStatement(
            ts.factory.createIdentifier(ts.transpile(script)),
        );

    export const block = (expression: ts.Expression) =>
        ts.factory.createBlock(
            [ts.factory.createExpressionStatement(expression)],
            true,
        );
}

File: packages/core/src/factories/StatementFactory.go
package factories

type statementFactory struct{}

var StatementFactory statementFactory

func (statementFactory) Mut(props VariableStatementProps) Statement {
	typeNode := props.Type
	if typeNode.Kind == "" && props.Initializer.Kind == "" {
		typeNode = TypeFactory.Keyword("any")
	}
	return Node("variableStatement", Field("flags", "let"), Field("name", props.Name), Field("type", typeNode), Field("initializer", props.Initializer))
}

func (statementFactory) Constant(props VariableStatementProps) Statement {
	typeNode := props.Type
	if typeNode.Kind == "" && props.Initializer.Kind == "" {
		typeNode = TypeFactory.Keyword("any")
	}
	return Node("variableStatement", Field("flags", "const"), Field("name", props.Name), Field("type", typeNode), Field("initializer", props.Initializer))
}

func (statementFactory) Entry(props EntryProps) Statement {
	return Node("variableDeclarationList", Field("flags", "const"), Field("binding", []string{props.Key, props.Value}))
}

func (statementFactory) Transpile(script string) Statement {
	return Node("expressionStatement", Field("expression", Identifier(script)))
}

func (statementFactory) Block(expression Expression) Statement {
	return Node("block", Field("statements", []Statement{Node("expressionStatement", Field("expression", expression))}), Field("multiline", true))
}
